sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c

```diff
@@ -0,0 +1,390 @@
+            gnrc_pktbuf_release(pkt);
+            return NULL;
+        }
+        memcpy(opt_snip->data, &p2p_ext->addr_vec[i-1], addr_len);
+        pkt = opt_snip;
+    }
+
+    if ((opt_snip = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_rpl_p2p_opt_rdo_t),
+                                    GNRC_NETTYPE_UNDEF)) == NULL) {
+        DEBUG("RPL: BUILD RDO - no space left in packet buffer\n");
+        gnrc_pktbuf_release(pkt);
+        return NULL;
+    }
+    rdo = opt_snip->data;
+    rdo->type = GNRC_RPL_P2P_OPT_RDO;
+    rdo->length = GNRC_RPL_P2P_RDO_LEN +
+                  p2p_ext->addr_numof * (sizeof(ipv6_addr_t) - p2p_ext->compr);
+    rdo->compr_flags = (p2p_ext->reply << GNRC_RPL_P2P_RDO_FLAGS_REPLY) |
+                        (p2p_ext->hop_by_hop << GNRC_RPL_P2P_RDO_FLAGS_HbH)
+                        | ((p2p_ext->routes_numof & 0x3) << GNRC_RPL_P2P_RDO_FLAGS_ADDRNUM)
+                        | (p2p_ext->compr & GNRC_RPL_P2P_RDO_FLAGS_COMPR);
+    rdo->lmn = ((p2p_ext->lifetime_enc & 0x3) << GNRC_RPL_P2P_RDO_FLAGS_LIFETIME) |
+                (p2p_ext->maxrank & GNRC_RPL_P2P_RDO_FLAGS_MAXRANK);
+    rdo->target = p2p_ext->target;
+    return opt_snip;
+}
+
+void gnrc_rpl_p2p_rdo_parse(gnrc_rpl_p2p_opt_rdo_t *rdo, gnrc_rpl_dodag_t *dodag)
+{
+    DEBUG("RPL: Route Discovery DIO option parsed\n");
+    gnrc_rpl_p2p_ext_t *p2p_ext = gnrc_rpl_p2p_ext_get(dodag);
+
+    uint8_t addr_num = (rdo->length - GNRC_RPL_P2P_RDO_LEN)
+                        / (sizeof(ipv6_addr_t) - p2p_ext->compr);
+    if (addr_num >= GNRC_RPL_P2P_ADDR_VEC_NUMOF) {
+        DEBUG("RPL: cannot parse RDO - too many hops\n");
+        return;
+    }
+
+    p2p_ext->for_me = (gnrc_ipv6_netif_find_by_addr(NULL, &rdo->target) != KERNEL_PID_UNDEF);
+
+    p2p_ext->reply = (rdo->compr_flags & (1 << GNRC_RPL_P2P_RDO_FLAGS_REPLY))
+                      >> GNRC_RPL_P2P_RDO_FLAGS_REPLY;
+    p2p_ext->hop_by_hop = (rdo->compr_flags & (1 << GNRC_RPL_P2P_RDO_FLAGS_HbH))
+                           >> GNRC_RPL_P2P_RDO_FLAGS_HbH;
+    p2p_ext->routes_numof = (rdo->compr_flags & (0x3 << GNRC_RPL_P2P_RDO_FLAGS_ADDRNUM))
+                             >> GNRC_RPL_P2P_RDO_FLAGS_ADDRNUM;
+    p2p_ext->compr = rdo->compr_flags & GNRC_RPL_P2P_RDO_FLAGS_COMPR;
+    p2p_ext->lifetime_enc = (rdo->lmn & (0x3 << GNRC_RPL_P2P_RDO_FLAGS_LIFETIME))
+                             >> GNRC_RPL_P2P_RDO_FLAGS_LIFETIME;
+    if (p2p_ext->lifetime_sec == INT8_MIN) {
+        p2p_ext->lifetime_sec = gnrc_rpl_p2p_lifetime_lookup[p2p_ext->lifetime_enc];
+    }
+    p2p_ext->maxrank = rdo->lmn & GNRC_RPL_P2P_RDO_FLAGS_MAXRANK;
+    p2p_ext->target = rdo->target;
+
+    memset(&p2p_ext->addr_vec, 0, sizeof(ipv6_addr_t) * GNRC_RPL_P2P_ADDR_VEC_NUMOF);
+    p2p_ext->addr_numof = 0;
+    uint8_t *tmp = (uint8_t *) (rdo + 1);
+    uint8_t *addr = NULL;
+    uint8_t addr_len = sizeof(ipv6_addr_t) - p2p_ext->compr;
+    uint8_t i = 0;
+    for (i = 0; i < addr_num; i++) {
```

Comment:
hm, I tried that approach but it was not so trivial (: Reverted the change for now.